pkg/controller/keyvault/spc/ingress.go

```diff
@@ -0,0 +1,155 @@
+package spc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"strings"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	netv1 "k8s.io/api/networking/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	secv1 "sigs.k8s.io/secrets-store-csi-driver/apis/v1"
+)
+
+var ingressSecretProviderControllerName = controllername.New("keyvault", "ingress", "secret", "provider")
+
+func NewIngressSecretProviderClassReconciler(manager ctrl.Manager, conf *config.Config, ingressManager util.IngressManager) error {
+	metrics.InitControllerMetrics(ingressSecretProviderControllerName)
+	if conf.DisableKeyvault {
+		return nil
+	}
+
+	spcReconciler := &secretProviderClassReconciler[*netv1.Ingress]{
+		name: ingressSecretProviderControllerName,
+		toSpcOpts: func(_ context.Context, _ client.Client, ing *netv1.Ingress) iter.Seq2[spcOpts, error] {
+			return ingressToSpcOpts(conf, ing, ingressManager)
+		},
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	}
+
+	return ingressSecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&netv1.Ingress{}).
+			Owns(&secv1.SecretProviderClass{}),
+		manager.GetLogger(),
+	).Complete(spcReconciler)
+}
+
+func ingressToSpcOpts(conf *config.Config, ing *netv1.Ingress, ingressManager util.IngressManager) iter.Seq2[spcOpts, error] {
+	return func(yield func(spcOpts, error) bool) {
+		if conf == nil {
+			yield(spcOpts{}, errors.New("config is nil"))
```

Comment:
out of curiosity - why not use a pointer here? feels like it would be slightly more efficient, and nil handling could be a little more concrete right?